core/kernel/wait_queue.c

```diff
@@ -69,24 +75,51 @@ void wq_wait_init_condvar(struct wait_queue *wq, struct wait_queue_elem *wqe,
 	cpu_spin_unlock_xrestore(&wq_spin_lock, old_itr_status);
 }
 
-void wq_wait_final(struct wait_queue *wq, struct wait_queue_elem *wqe,
-		   const void *sync_obj, const char *fname, int lineno)
+static TEE_Result wq_wait_final_helper(struct wait_queue *wq,
+				       struct wait_queue_elem *wqe,
+				       uint32_t timeout_ms,
+				       const void *sync_obj, const char *fname,
+				       int lineno)
 {
-	uint32_t old_itr_status;
-	unsigned done;
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t old_itr_status = 0;
+	bool done = false;
 
 	do {
-		do_notif(notif_wait, wqe->handle,
-			 "sleep", sync_obj, fname, lineno);
+		res = do_notif(true, wqe->handle, timeout_ms,
+			       "sleep", sync_obj, fname, lineno);
 
 		old_itr_status = cpu_spin_lock_xsave(&wq_spin_lock);
 
+		if (wqe->done)
+			res = TEE_SUCCESS;
+		else
+			res = TEE_ERROR_TIMEOUT;
+		wqe->done = true;
```

Comment:
Do this really work? One waiting for a event may be woken but it's resource not yet ready hence going back for another roudnin this loop. Something looks odd to me.

Reply:
Reading back https://github.com/OP-TEE/optee_os/pull/6641#discussion_r1548092078 I think I miss understood `wqe->done` value. 

If the sequence here is fine, then I would suggest to simplify this:
```c
{
	uint32_t old_itr_status = 0;

	do_notif(true, wqe->handle, timeout_ms, "sleep", sync_obj, fname,
		 lineno);

	old_itr_status = cpu_spin_lock_xsave(&wq_spin_lock);
	SLIST_REMOVE(wq, wqe, wait_queue_elem, link);
	cpu_spin_unlock_xrestore(&wq_spin_lock, old_itr_status);

	if (wqe->done)
		return TEE_SUCCESS;
	else
		return TEE_ERROR_TIMEOUT;
}
```
I even think `wq_wait_final_helper()` does not need a return value. Caller could check `wqe->done` to know whether the notif waiting process timed out or not. That said, an explicit return value may be convenient.

Reply:
I was about to suggest the same. I think we can simplify further. We don't need a separate `wq_wait_final_timeout()`, we could add the `timeout_ms` parameter and return value to `wq_wait_final()`.

Reply:
> I even think `wq_wait_final_helper()` does not need a return value. Caller could check `wqe->done` to know whether the notif waiting process timed out or not. That said, an explicit return value may be convenient.

The return value of ```wq_wait_final_helper``` will propagated up to ```condvar_wait_timeout```.
The caller can check return the value from ```condvar_wait_timeout```. However, since the caller cannot get ```wqe```, they cannot check ```wqe->done```. Therefore, the return value of ```wq_wait_final_helper``` is required. Is my understanding correct?

Reply:
> I was about to suggest the same. I think we can simplify further. We don't need a separate `wq_wait_final_timeout()`, we could add the `timeout_ms` parameter and return value to `wq_wait_final()`.

Ok, I can extend `wq_wait_final()` with parameter `timeout_ms`.